Sources/System/FilePath/FilePathSyntax.swift

```diff
@@ -0,0 +1,353 @@
+  /// C++ was a little misguided here after looking more at C#. I think
+  /// a better alternative would be to actually identify and work with
+  /// the root when on Windows, more akin to Rust's `Prefix` type.
+
+  /// Gets or sets the relative portion of the path (everything after root)
+  ///
+  /// Examples:
+  /// * Unix:
+  ///   * `/foo/bar => foo/bar`
+  ///   * `foo/bar => foo/bar`
+  ///   * `/ => ""`
+  /// * Windows:
+  ///   * `C:\foo\bar => foo\bar`
+  ///   * `foo\bar => foo\bar`
+  ///   * `\\?\UNC\server\share\file => file`
+  ///   * `\ => ""`
+  ///
+  /// REVIEW NOTE: Not optional, can return empty paths. Empty paths
+  /// are relative (but are not roots), hence non-optional.
+  public var relativePath: FilePath {
+    get {
+      let (_, relIdx) = _parseRoot()
+      return FilePath(slice: storage[relIdx...])
+    }
+    set { fatalError() }
+  }
+
+  /// Returns the final component of the path. If the path is a normal file, this
+  /// will be the file name. If the path is a directory, this is the directory name.
+  /// Returns `nil` if the path is empty or only contains a root
+  ///
+  /// REVIEW NOTE: We follow  `basename` semantics, unlike `file_name` in
+  /// C++ which treats trailing `/` specially, or in Rust which returns nil for `..`
+  ///
+  /// REVIEW NOTE: shell and C basename will return `/` if the path is `/`. I don't
+  /// think we want to, it seems clearer that this only is in the relative portion of the path,
+  /// and doing so would make the setter weird.
+  public var basename: FilePath.Component? {
+    get {
+      let (rootEnd, _) = _parseRoot()
+      guard let base = self.components.last, base.slice.startIndex >= rootEnd else {
+        return nil
+      }
+      return base
+    }
+    set {
+      // TODO: this is assuming that pop correctly preserves root
+      _ = self.pop()
+      if let c = newValue { self.append(FilePath(c)) }
+    }
+  }
+
+  /// Returns the path up to but not including the `basename`.
+  ///
+  /// If the path only contains a root, returns `self`.
+  /// If the path has no root and only includes a single component,
+  /// returns an empty FilePath.
+  ///
+  /// REVIEW NOTE: Shell returns `.` if there is no dirname, while we return the empty path.
+  /// REVIEW NOTE: I removed the setter. It seems sketchy and its use indicates programmer error.
+  public var dirname: FilePath {
+    get {
+      // No basename means path is root or empty (which is the dirname)
+      if self.basename == nil { return self }
+      return FilePath(self.components.dropLast())
+    }
+  }
+
+  /// The extension of the file or directoy last component.
+  ///
+  /// If `basename` is `nil` or one of the special path components
+  /// `.` or `..`, `get` returns `nil` and `set` does nothing.
+  ///
+  /// If `basename` does not contain a `.` anywhere, or only
+  /// at the start, `get` returns `nil` and `set` will append a
+  /// `.` and `newValue` to `basename`
```

Comment:
I think there are additional rules on some platforms, such as not treating extensions consisting only of digits as extensions (e.g. "Foo 1.0" doesn't have the extension "0"), but I'm guessing that might be a bit too specialized for this API.

Reply:
That's an interesting point. Would this be a platform-specific consideration or a file-system-specific consideration or application-specific?

Reply:
It would be a platform-specific consideration.  I looked to see if I could find anything more specific about macOS, which is the platform I was thinking of with the no-digits consideration, but I couldn't find anything relevant so I may be misremembering.  This was something that was discussed a lot around the time of Mac OS X 10.0, when the Classic MacOS semantics were being reconciled with the Unix-based conventions of OPENSTEP.  But I think I'm either misremembering or things have changed, since it's just that the `.` can't be the first or last character of the last path component.